client/client-multi/client-c/src/kaa/kaa_notification_manager.c

```diff
@@ -0,0 +1,1035 @@
+                kaa_list_destroy(notification_listeners, &kaa_data_destroy);
+                kaa_list_remove_at(&self->optional_listeners, optional_listener_node, &kaa_data_destroy);
+            }
+            outdated_topics = kaa_list_next(outdated_topics);
+        }
+    }
+    if (self->topics) {
+        kaa_list_destroy(self->topics, &kaa_topics_destroy);
+    }
+    self->topics = topics;
+    return kaa_notify_topic_update_subscribers(self, topics);
+}
+
+kaa_error_t kaa_notification_received(kaa_notification_manager_t *self, kaa_notification_t *notification, uint64_t* topic_id)
+{
+    KAA_RETURN_IF_NIL2(self, notification, KAA_ERR_BADDATA);
+    kaa_topic_t* topic = NULL;
+    kaa_error_t err = kaa_find_topic(self, &topic, topic_id);
+    if (err) {
+        KAA_LOG_WARN(self->logger, KAA_ERR_NOT_FOUND, "Unknown notification received(topic id = %lu) ", *topic_id);
+    }
+    if (kaa_notify_optional_notification_subscribers(self, topic_id, notification)) {
+        kaa_notify_mandatory_notification_subscribers(self,topic_id, notification);
+    }
+    return KAA_ERR_NONE;
+}
+kaa_error_t kaa_notification_manager_handle_server_sync(kaa_notification_manager_t *self
+                                                      , kaa_platform_message_reader_t *reader
+                                                      , uint32_t extension_length)
+{
+    KAA_RETURN_IF_NIL2(self, reader, KAA_ERR_BADPARAM);
+
+    kaa_list_destroy(self->subscriptions, &kaa_data_destroy);
+    kaa_list_destroy(self->uids, &kaa_uids_destroy);
+    self->uids = NULL;
+    self->subscriptions = NULL;
+
+    kaa_error_t err = KAA_ERR_NONE;
+    if (extension_length > 0) {
+        self->status->notification_seq_n = KAA_NTOHL(*((uint32_t *) reader->current));
+        reader->current += sizeof(uint32_t);
+        extension_length -= sizeof(uint32_t);
+        uint8_t field_id = 0;
+        while (extension_length > 0) {
+            field_id = *((uint8_t *) reader->current);
+            reader->current += sizeof(uint16_t);
+            extension_length -= sizeof(uint16_t);
+            switch (field_id) {
+            case NOTIFICATIONS: {
+                kaa_notification_t *notification = NULL;
+                uint16_t uid_length = 0;
+                uint64_t topic_id = 0;
+                kaa_topic_t *topic_found = NULL;
+                uint32_t seq_number = 0;
+                uint32_t notification_size = 0;
+                uint16_t notifications_count = KAA_NTOHS(*((uint16_t *) reader->current));
+                reader->current += sizeof(uint16_t);
+                extension_length -= sizeof(uint16_t);
+                KAA_LOG_TRACE(self->logger, KAA_ERR_NONE, "Notifications count is %u", notifications_count);
+                if (notifications_count) {
+                    while (notifications_count--) {
```

Comment:
I think "notifications_count-- > 0" will be better, because at least we don't need two if-s in which the notifications_count value is checked.